src/java/org/apache/cassandra/index/sai/disk/vector/CompressedVectorRowIdIterator.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.disk.vector;
+
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.function.IntFunction;
+
+import io.github.jbellis.jvector.vector.VectorSimilarityFunction;
+import org.apache.cassandra.utils.CloseableIterator;
+
+/**
+ * An iterator over {@link ScoredRowId} that lazily orders a {@link PriorityQueue} of approximately scored rows.
+ * The class works by:
+ * 1. Taking a priority queue of {@link ApproximateScore} ordered by approximate similarity score.
+ * 2. Polling the topK results from the approximate score queue
+ * 3. Materialize and score the full resolution vector for each row from step 2.
+ * 4. Add to the full resolution row {@link ScoredRowId} to the priority queue.
+ * 5. Return rows from the {@link ScoredRowId} priority queue until the limit is reached.
+ * 6. When the limit is reached, if the {@link ApproximateScore} queue is not empty, get the next (topK - limit) rows
+ *    from the approximate score queue and repeat 3-6 until the {@link ApproximateScore} queue is empty.
+ * 7. Return the remaining rows from the {@link ScoredRowId} queue.
+ */
+public class CompressedVectorRowIdIterator implements CloseableIterator<ScoredRowId>
+{
+    public static class ApproximateScore implements Comparable<ApproximateScore>
+    {
+        private final int rowId;
+        private final int ordinal;
+        private final float appoximateScore;
+
+        public ApproximateScore(int rowId, int ordinal, float appoximateScore)
+        {
+            this.rowId = rowId;
+            this.ordinal = ordinal;
+            this.appoximateScore = appoximateScore;
+        }
+
+        @Override
+        public int compareTo(ApproximateScore o)
+        {
+            // We always want scores in descending order
```

Comment:
this is an ascending compareTo, is the comment or the code wrong?

Reply:
Which part is ascending? `approxScoreA.compareTo(approxScoreB)` results in `Float.compare(approxScoreB.appoximateScore, appoximateScore)`, which is descending. If you'd prefer not hiding the implementation within the class, I can put it in the PQ's constructor like I've seen elsewhere.

Reply:
all of the default JDK compare functions are ascending

(meaning that if you sort with them, you get the smallest results at the start)

this includes Float.compare
```
     * @return  the value {@code 0} if {@code f1} is
     *          numerically equal to {@code f2}; a value less than
     *          {@code 0} if {@code f1} is numerically less than
     *          {@code f2}; and a value greater than {@code 0}
     *          if {@code f1} is numerically greater than
     *          {@code f2}.
     * @since 1.4
     */
    public static int compare(float f1, float f2) {
 ```

Reply:
But because we flip the objects so that the left parameter is `o.appoximateScore`, it becomes inverted. We do this in PQs by passing `(a, b) -> Float.compare(b.score, a.score)`.

Reply:
Yep, I simply didn't notice the parameter order changed.